applications/StructuralMechanicsApplication/custom_elements/prestress_membrane_element.cpp

```diff
@@ -1132,142 +1135,552 @@ void PrestressMembraneElement::CalculateMembraneElasticityTensor(
+    CrossProduct(normal_projection_plane, GlobalPrestressAxis1, GlobalPrestressAxis3);
+    array_1d<double, 3> T1, T2, T3;
+    CrossProduct(T1, normal_projection_plane, E3);
+    T1 /= norm_2(T1);
+    CrossProduct(T2,E3,T1);
+    T2 /= norm_2(T2);
+    T3 = E3;
+    
 
-    array_1d<double, 3> t1; // T1 vector
-    array_1d<double, 3> t2; // T2 vector
-    array_1d<double, 3> t3; // T3 vector
-    array_1d<double, 3> a;  // A vector for the plane definition
-    array_1d<double, 3> b;  // B vector for the plane definition
-    boost::numeric::ublas::bounded_matrix<double, 3, 3> Tm; // Transformation matrix
 
-    // Hardcode the direction vector a, b
-    a(0) = 1.0;
-    a(1) = 0.0;
-    a(2) = 0.0;
+    // Transform prestresses in the local cartesian cosy in reference configuration
+    bounded_matrix<double,3,3> Origin = ZeroMatrix(3,3);
+    bounded_matrix<double,3,3> Target = ZeroMatrix(3,3);
+    bounded_matrix<double,3,3> Tensor = ZeroMatrix(3,3);
 
-    b(0) = 0.0;
-    b(1) = 1.0;
-    b(2) = 0.0;
+    for (int i=0;i<3;i++){
+        Origin(i,0) = T1(i);
+        Origin(i,1) = T2(i);
+        Origin(i,2) = T3(i);
+        Target(i,0) = E1(i);
+        Target(i,1) = E2(i);
+        Target(i,2) = E3(i);        
+    }
 
-    //array_1d<double, 3> g1;
-    //array_1d<double, 3> g2;
-    array_1d<double, 3> g1_ref;
-    array_1d<double, 3> g2_ref;
-    //array_1d<double, 3> g3;
-    //array_1d<double, 3> gab;
-    array_1d<double, 3> g_con_1_ref;
-    array_1d<double, 3> g_con_2_ref;
-    
-    //KRATOS_WATCH(g1);
+    Tensor.clear();
+    Tensor(0,0) = this->GetValue(MEMBRANE_PRESTRESS)(0,PointNumber);
+    Tensor(1,1) = this->GetValue(MEMBRANE_PRESTRESS)(1,PointNumber);
+    Tensor(1,0) = this->GetValue(MEMBRANE_PRESTRESS)(2,PointNumber);
+    Tensor(0,1) = this->GetValue(MEMBRANE_PRESTRESS)(2,PointNumber);
 
-    // contravariant metric gab_con and base vectors g_con
-    double invdetGab = 1.0 / (gab[0] * gab[1] - gab[2] * gab[2]);
-    double gab_con11 = invdetGab*gab[1];
-    double gab_con12 = -invdetGab*gab[2];
-    double gab_con22 = -invdetGab*gab[0];
+    // Transformation Stress Tensor
+    StructuralMechanicsMathUtilities::TensorTransformation<3>(Origin,Origin,Target,Target,Tensor);
 
-    array_1d<double, 3> g_con_1 = g1*gab_con11 + g2*gab_con12;
-    array_1d<double, 3> g_con_2 = g1*gab_con12 + g2*gab_con22;
+    // store prestress values in the elemental data
+    Matrix& m = GetValue(MEMBRANE_PRESTRESS);
+    m(0,PointNumber) = Tensor(0,0);
+    m(1,PointNumber) = Tensor(1,1);
+    m(2,PointNumber) = Tensor(1,0);
+}
 
-    // local cartesian coordinates
-    double lg1 = norm_2(g1);
-    array_1d<double, 3> e1 = g1 / lg1;
-    double lg_con2 = norm_2(g_con_2);
-    array_1d<double, 3> e2 = g_con_2 / lg_con2;
+// in each solution step: update reference configuration, update prestress (in the anisotropic case)
+void PrestressMembraneElement::InitializeSolutionStep(ProcessInfo& rCurrentProcessInfo){
```

Comment:
This function is also very long, can you please devide it into smaller ones, if possible?

Reply:
done